File: packages/native-android/android/src/main/java/com/openota/runtime/BundleStorage.kt
package com.openota.runtime

import android.content.Context
import java.io.File
import java.nio.file.Files

/**
 * All filesystem access for the runtime goes through here. Nothing outside this class knows the
 * literal directory names, and nothing outside this class is allowed to build a [File] from a raw
 * string without going through [resolveWithinRoot] — that single chokepoint is what makes "never
 * write outside OpenOTA" and "reject path traversal / symlinks" enforceable in one place instead
 * of scattered across every caller.
 *
 * Android has no literal "Documents" directory the way iOS does; `context.filesDir` is its closest
 * equivalent (private, app-scoped, persists across app updates, wiped on uninstall), so the
 * `OpenOTA/` root is rooted there rather than external storage.
 */
class BundleStorage(context: Context) {
    private val appContext = context.applicationContext

    val rootDir: File = File(appContext.filesDir, "OpenOTA").apply { mkdirs() }
    val currentDir: File = File(rootDir, "current").apply { mkdirs() }
    val rollbackDir: File = File(rootDir, "rollback").apply { mkdirs() }
    val downloadsDir: File = File(rootDir, "downloads").apply { mkdirs() }
    val cacheDir: File = File(rootDir, "cache").apply { mkdirs() }
    val tmpDir: File = File(rootDir, "tmp").apply { mkdirs() }
    val logsDir: File = File(rootDir, "logs").apply { mkdirs() }
    val rootManifestFile: File = File(rootDir, "manifest.json")

    fun readRootManifest(): RuntimeManifest =
        if (rootManifestFile.exists()) {
            runCatching { RuntimeManifest.parse(rootManifestFile.readText()) }.getOrDefault(RuntimeManifest.EMPTY)
        } else {
            RuntimeManifest.EMPTY
        }

    fun writeRootManifest(manifest: RuntimeManifest) {
        val tmpFile = File(tmpDir, "manifest.json.tmp")
        tmpFile.writeText(manifest.toJson())
        if (!tmpFile.renameTo(rootManifestFile)) {
            rootManifestFile.writeText(manifest.toJson())
            tmpFile.delete()
        }
    }

    /**
     * Resolves an externally-supplied path (e.g. the argument to `setBundlePath`) against a
     * permitted parent directory, rejecting anything that isn't a plain relative descendant.
     * This is the runtime's core defense against path traversal and symlink escape: absolute
     * paths, `..` segments, and symlinked components are all rejected before a single byte of
     * the candidate bundle is read.
     */
    fun resolveWithinRoot(candidatePath: String): File {
        val candidate = File(candidatePath)

        if (candidate.isAbsolute && !candidate.canonicalPath.startsWith(rootDir.canonicalPath + File.separator) &&
            candidate.canonicalPath != rootDir.canonicalPath
        ) {
            throw PathSecurityException("Path \"$candidatePath\" is outside the OpenOTA sandbox")
        }

        val resolved = if (candidate.isAbsolute) candidate else File(rootDir, candidatePath)
        val canonical = resolved.canonicalFile

        if (!canonical.path.startsWith(rootDir.canonicalPath + File.separator) &&
            canonical.path != rootDir.canonicalPath
        ) {
            throw PathSecurityException("Path \"$candidatePath\" escapes the OpenOTA root directory")
        }

        return canonical
    }

    /** Rejects a directory tree containing any symbolic link, walked without following links. */
    fun assertNoSymlinks(dir: File) {
        if (!dir.exists()) return

        dir.walkTopDown().forEach { file ->
            if (Files.isSymbolicLink(file.toPath())) {
                throw PathSecurityException("Symbolic link detected at \"${file.path}\"")
            }
        }
    }

    /** Atomically (same-volume rename) replaces [destination] with [source], deleting any prior contents. */
    fun atomicReplace(source: File, destination: File) {
        deleteRecursively(destination)
        destination.parentFile?.mkdirs()

        if (!source.renameTo(destination)) {
            source.copyRecursively(destination, overwrite = true)
            deleteRecursively(source)
        }
    }

    fun deleteRecursively(file: File) {
        if (file.exists()) {
            file.deleteRecursively()
        }
    }

    fun bundleFile(packageDir: File, bundleFileName: String): File = File(File(packageDir, "bundle"), bundleFileName)

    fun assetsDir(packageDir: File): File = File(packageDir, "assets")

    fun manifestFile(packageDir: File): File = File(packageDir, "manifest.json")
}

File: packages/native-ios/ios/BundleStorage.swift
import Foundation

/// All filesystem access for the runtime goes through here — mirrors `BundleStorage.kt`. Nothing
/// outside this class builds a path from a raw string without going through `resolveWithinRoot`,
/// which is the single chokepoint enforcing "never write outside OpenOTA" and "reject path
/// traversal / symlinks".
///
/// iOS's equivalent of Android's `context.filesDir` is `Application Support` — private,
/// app-scoped, persists across updates, excluded from iCloud backup once marked, and (per Apple's
/// guidance) the correct home for app-managed data that isn't user-visible documents. Never
/// Documents/Caches for the *active* bundle: Caches can be purged by the OS under disk pressure at
/// any time, which would silently break app launch.
final class BundleStorage {
    let rootDir: URL
    let currentDir: URL
    let rollbackDir: URL
    let downloadsDir: URL
    let cacheDir: URL
    let tmpDir: URL
    let logsDir: URL
    let stateFile: URL

    private let fm = FileManager.default

    /// `root` is injectable so tests run against an isolated temp directory instead of the real
    /// Application Support path (spec requirement: hermetic tests).
    init(root: URL? = nil) {
        let base: URL
        if let root {
            base = root
        } else {
            let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            base = appSupport.appendingPathComponent("OpenOTA", isDirectory: true)
        }
        rootDir = base
        currentDir = base.appendingPathComponent("current", isDirectory: true)
        rollbackDir = base.appendingPathComponent("rollback", isDirectory: true)
        downloadsDir = base.appendingPathComponent("downloads", isDirectory: true)
        cacheDir = base.appendingPathComponent("cache", isDirectory: true)
        tmpDir = base.appendingPathComponent("tmp", isDirectory: true)
        logsDir = base.appendingPathComponent("logs", isDirectory: true)
        stateFile = base.appendingPathComponent("state.json")

        for dir in [rootDir, currentDir, rollbackDir, downloadsDir, cacheDir, tmpDir, logsDir] {
            try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        // Exclude the whole OpenOTA tree from iCloud/iTunes backup — OTA bundles are re-downloadable
        // and shouldn't bloat user backups or leak into them.
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        var mutableRoot = rootDir
        try? mutableRoot.setResourceValues(values)
    }

    func readState() -> RuntimeManifest {
        guard let data = try? Data(contentsOf: stateFile) else { return .empty }
        return RuntimeManifest.parse(data)
    }

    /// Temp-write + rename so a crash mid-write never leaves a torn/partial `state.json` readable
    /// as valid JSON.
    func writeState(_ manifest: RuntimeManifest) throws {
        let data = try manifest.toJSON()
        let tmpFile = tmpDir.appendingPathComponent("state.json.tmp")
        try data.write(to: tmpFile, options: .atomic)
        _ = try? fm.removeItem(at: stateFile)
        try fm.moveItem(at: tmpFile, to: stateFile)
    }

    /// Resolves an externally-supplied path against the OpenOTA sandbox, rejecting anything that
    /// isn't a plain descendant. Defends against path traversal and symlink escape before a single
    /// byte of the candidate bundle is read.
    func resolveWithinRoot(_ candidatePath: String) throws -> URL {
        var path = candidatePath
        if path.hasPrefix("file://"), let url = URL(string: path) {
            path = url.path
        }

        let candidate = URL(fileURLWithPath: path)
        let resolved: URL = candidate.path.hasPrefix("/") ? candidate : rootDir.appendingPathComponent(path)

        let standardizedRoot = rootDir.standardizedFileURL.resolvingSymlinksInPath().path
        let standardizedCandidate = resolved.standardizedFileURL.resolvingSymlinksInPath().path

        guard standardizedCandidate == standardizedRoot || standardizedCandidate.hasPrefix(standardizedRoot + "/") else {
            throw OpenOTAException.pathSecurityError("Path \"\(candidatePath)\" escapes the OpenOTA root directory")
        }
        return URL(fileURLWithPath: standardizedCandidate)
    }

    /// Rejects a directory tree containing any symbolic link, without following links.
    func assertNoSymlinks(_ dir: URL) throws {
        guard fm.fileExists(atPath: dir.path) else { return }
        guard let enumerator = fm.enumerator(
            at: dir,
            includingPropertiesForKeys: [.isSymbolicLinkKey],
            options: [.producesRelativePathURLs]
        ) else { return }

        for case let fileURL as URL in enumerator {
            let values = try fileURL.resourceValues(forKeys: [.isSymbolicLinkKey])
            if values.isSymbolicLink == true {
                throw OpenOTAException.pathSecurityError("Symbolic link detected at \"\(fileURL.path)\"")
            }
        }
    }

    /// Atomically (same-volume rename) replaces `destination` with `source`, deleting any prior
    /// contents first.
    func atomicReplace(source: URL, destination: URL) throws {
        deleteRecursively(destination)
        try? fm.createDirectory(at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)

        do {
            try fm.moveItem(at: source, to: destination)
        } catch {
            // Cross-volume fallback: copy then delete the source.
            try fm.copyItem(at: source, to: destination)
            deleteRecursively(source)
        }
    }

    func deleteRecursively(_ url: URL) {
        try? fm.removeItem(at: url)
    }

    func bundleFile(_ packageDir: URL, bundleFileName: String) -> URL {
        packageDir.appendingPathComponent("bundle", isDirectory: true).appendingPathComponent(bundleFileName)
    }

    func assetsDir(_ packageDir: URL) -> URL {
        packageDir.appendingPathComponent("assets", isDirectory: true)
    }

    func manifestFile(_ packageDir: URL) -> URL {
        packageDir.appendingPathComponent("manifest.json")
    }

    func fileExists(_ url: URL, isDirectory: Bool? = nil) -> Bool {
        var isDir: ObjCBool = false
        let exists = fm.fileExists(atPath: url.path, isDirectory: &isDir)
        if let isDirectory {
            return exists && isDir.boolValue == isDirectory
        }
        return exists
    }

    func copyRecursively(from: URL, to: URL) throws {
        deleteRecursively(to)
        try fm.copyItem(at: from, to: to)
    }
}
